### PR TITLE
Refactors throwing

### LINKED ISSUE #20077: Thrown objects work weird in melee range, unable to embed
**Issue Description**:
Certain thrown objects such as throwing stars or spears cannot embed when thrown from a tile adjacent to your target.

**What did you expect to happen**: 
Certain items are unable to embed or cause knockback when thrown in melee range.

**What happened instead**:
Spears, throwing stars, gar mesons and presumably anything able to embed will oddly pass through the target, damage them but fail to embed despite a chat log indicating it embeded. 

**Why is this bad/What are the consequences**:
Bugs are bad and the game shouldn't say something embeded when it in fact did not

**Steps to reproduce the problem**:
Stand next to a human,  throw a spear or throwing star while standing next to them. Note that this bug only occurs if there is NOT a wall or otherwise impassable object opposite of the target and the direction of which you throw from.

**When did the problem start happening**:
Noticed today

**Extra information**:
Objects that do not embed behave more normally with the only oddity being that some objects will fall at the targets feet while others will fall at your own feet

Screenshot_24
Screenshot_25

### LINKED ISSUE #22009: Missing Embeded Objects
**Issue Description**:
Embeded objects disappear, but still leave the icon in the top right.

**What did you expect to happen**: 
Expected a embed object to exist

**What happened instead**:
Embed object didn't exist

**Why is this bad/What are the consequences**:
Left a embed icon in the top right that had no reason to be there

**Steps to reproduce the problem**:
Not sure if it can be reproduced, there was a gravitational anomaly, and myself and at least one other had embed objects. We walked to medbay to get them taken out but medics said there was nothing there. Didn't show up in scans and there was no embed removal surgery option.

**When did the problem start happening**:

**Extra information**:
Removing the limb and reattaching the limb made the icon go away.

### PR DESCRIPTION
**Pull request by chuga-git**
## What Does This PR Do

Fixes #20077
Fixes #22009

Port of https://github.com/tgstation/tgstation/pull/60604 and later bugfixes.

Our old version of SSthrowing had a lot of crust and lacked proper GC handling. #25921 was made in response to a case where catching the thrown grenade would cause it to both deposit into your hand and land one tile behind you. The rough sequence of events I narrowed the cause down to is as follows:
1. The boomerang component calls `throw_at()` on the rebound, likely choosing a turf directly behind the mob.
2. The next `tick()` happens and the grenade is moved one turf in the direction of the mob.
3. Moving onto this turf triggers `Bump` against the mob, which in turn calls `throwing.hit_atom(atom)`, which then calls `finalize()`. `finalize()` triggers the mob's `hit_by`, which eventually results in the grenade being `forceMove`d into its contents (and hands).  This *should* be the end of the story, but we're actually still in the `while` loop of the `tick` that caused the `Bump`. The `finalize()` was a side effect of the `AM.Move()` call, not a terminating branch of the loop.
4. So, it loops one more time despite the item having its `throwing` reference cleared (no clue, man). This causes *another* call to `AM.Move()`, from the mob's `loc` to the `target_turf` directly behind it. Note that this does not actually remove it from your hands.
5. I become very distressed about the fact that this has been a thing for 7 years, and spot that TG has solved this issue already.

The majority of the changes have to do with making sure that when `finalize()` gets called, it properly cleans up after itself. This means that things being `forceMove`d on hit no longer have a chance of bugging out, which has caused untold pain and suffering with embeds, bolas, and catching.

The most noticeable impact on gameplay is that mobs who are knocked down / floored can now be hit by thrown items given that you're aiming at them.
## Why It's Good For The Game

SSthrowing is now slightly less of an eldritch horror full of undefined behavior. 
## Testing

Works fine in initial testing, needs much more.

## Changelog
:cl: Chuga
add: You can now throw things at mobs lying down on the floor.
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/SSthrowing.dm [hunk 1]
```diff
@@ -93,6 +118,48 @@ SUBSYSTEM_DEF(throwing)
 	/// Will thrownthing datum actually block movement? this might be FALSE with some circumstances even if var/should_block_movement is TRUE. This variable change automatically during the throw
 	var/block_movement = TRUE
 
+/datum/thrownthing/New(thrownthing, atom/target, init_dir, maxrange, speed, atom/thrower, diagonals_first, force, callback, dodgeable, block_movement, target_zone)
+	. = ..()
```
Comment (lewcc) [inline code/controllers/subsystem/SSthrowing.dm:122]: Is there actually a parent/behavior that this is calling?

File: code/controllers/subsystem/SSthrowing.dm [hunk 2]
```diff
@@ -93,6 +118,48 @@ SUBSYSTEM_DEF(throwing)
...
+	thrownthing.throwing = null
+	thrownthing = null
+	thrower_uid = null
+	initial_target_uid = null
+	callback = null
+	return ..()
+
+///Defines the datum behavior on the thrownthing's qdeletion event.
+/datum/thrownthing/proc/on_thrownthing_qdel(atom/movable/source, force)
+	SIGNAL_HANDLER
```
Comment (lewcc) [inline code/controllers/subsystem/SSthrowing.dm:153]: comment the sig

File: code/controllers/subsystem/SSthrowing.dm [hunk 3]
```diff
@@ -145,51 +222,55 @@ SUBSYSTEM_DEF(throwing)
...
+			return
 
 		dist_travelled++
 
-		if(dist_travelled > MAX_THROWING_DIST)
+		if(actual_target && !(actual_target.pass_flags_self & LETPASSTHROW) && actual_target.loc == AM.loc) // we crossed a movable with no density (e.g. a mouse or APC) we intend to hit anyway.
+			finalize(TRUE, actual_target)
+			return
+
+		if (dist_travelled > MAX_THROWING_DIST)
```
Comment (lewcc) [inline code/controllers/subsystem/SSthrowing.dm:236]: tg port moment
```suggestion
		if(dist_travelled > MAX_THROWING_DIST)
```

File: code/controllers/subsystem/SSthrowing.dm [hunk 4]
```diff
@@ -145,51 +222,55 @@ SUBSYSTEM_DEF(throwing)
...
 			finalize()
 			return
 
 /datum/thrownthing/proc/finalize(hit = FALSE, target = null)
-	set waitfor = 0
-	SSthrowing.processing -= thrownthing
+	set waitfor = FALSE
 	//done throwing, either because it hit something or it finished moving
+	if(!thrownthing)
```
Comment (lewcc) [inline code/controllers/subsystem/SSthrowing.dm:243]: why not check for qdeleted here

Comment (chuga-git) [reply code/controllers/subsystem/SSthrowing.dm:243]: `on_thrownthing_qdel()` will get triggered by COMSIG_PARENT_QDELETING before `finalize()` will be reached.

File: code/controllers/subsystem/SSthrowing.dm (hunk 1 above)
Comment (chuga-git) [reply code/controllers/subsystem/SSthrowing.dm:122]: Yes. `/atom/New(loc, ...)` in `atoms.dm` line 117.

Comment (chuga-git) [reply code/controllers/subsystem/SSthrowing.dm:122]: this is a cargo cult moment

**Review by Sirryan2002 — APPROVED** [review]
from just looking at the code, stuff looks fine. Edge cases might pop up in production but I have no idea lol.

**Comment by chuga-git** [comment]
> from just looking at the code, stuff looks fine. Edge cases might pop up in production but I have no idea lol.

Will need to keep a close eye on GC to make sure I didn't miss a patch. There were a few issues with thrownthing hard dels after the initial PR was released on TG, but the version I ported seems to be completely fixed.

File: code/controllers/subsystem/SSthrowing.dm (hunk 4 above)
Comment (lewcc) [reply code/controllers/subsystem/SSthrowing.dm:242]: Makes sense

**Review by lewcc — APPROVED** [review]
Yeah, LGTM. We can see if anything slips out.

**Pull request merged**